src/v0/destinations/marketo_bulk_upload/util.js

```diff
@@ -1,98 +1,343 @@
+const stats = require('../../../util/stats');
+const {
+  ABORTABLE_CODES,
+  THROTTLED_CODES,
+  POLL_ACTIVITY,
+  UPLOAD_FILE,
+  JOB_STATUS_ACTIVITY,
+  FETCH_ACCESS_TOKEN,
+  POLL_STATUS_ERR_MSG,
+  FILE_UPLOAD_ERR_MSG,
+  FETCH_FAILURE_JOB_STATUS_ERR_MSG,
+  FETCH_WARNING_JOB_STATUS_ERR_MSG,
+  ACCESS_TOKEN_FETCH_ERR_MSG,
+} = require('./config');
+const Cache = require('../../util/cache');
 
-const ABORTABLE_CODES = ['ENOTFOUND', 'ECONNREFUSED', 603, 605, 609, 610];
-const RETRYABLE_CODES = ['EADDRINUSE', 'ECONNRESET', 'ETIMEDOUT', 713, 601, 602, 604, 611];
-const THROTTLED_CODES = [502, 606, 607, 608, 615];
+const { AUTH_CACHE_TTL } = require('../../util/constant');
 
-const MARKETO_FILE_SIZE = 10485760;
-const MARKETO_FILE_PATH = `${__dirname}/uploadFile/marketo_bulkupload.csv`;
+const authCache = new Cache(AUTH_CACHE_TTL);
 
-const POLL_ACTIVITY = 'marketo_bulk_upload_polling';
-const UPLOAD_FILE = 'marketo_bulk_upload_upload_file';
-const JOB_STATUS_ACTIVITY = 'marketo_bulk_upload_get_job_status';
+const getMarketoFilePath = () =>
+  `${__dirname}/uploadFile/${Date.now()}_marketo_bulk_upload_${generateUUID()}.csv`;
 
-const getMarketoFilePath = () => MARKETO_FILE_PATH;
-// Fetch access token from client id and client secret
-// DOC: https://developers.marketo.com/rest-api/authentication/
-const getAccessToken = async (config) => {
+const getAccessTokenCacheKey = (config) => {
+  const { munchkinId, clientId, clientSecret } = config;
+  return `${munchkinId}-${clientId}-${clientSecret}`;
+};
+
+/**
+ * Handles common error responses returned from API calls.
+ * Checks the error code and throws the appropriate error object based on the code.
+ *
+ * @param {object} resp - The response object containing the error information.
+ * @param {string} OpErrorMessage - The error message to be used if the error code is not recognized.
+ * @param {string} OpActivity - The activity name for tracking purposes.
+ * @throws {AbortedError} - If the error code is abortable.
+ * @throws {ThrottledError} - If the error code is within the range of throttled codes.
+ * @throws {RetryableError} - If the error code is neither abortable nor throttled.
+ *
+ * @example
+ * const resp = {
+ *   response: {
+ *     errors: [
+ *       {
+ *         code: "1003",
+ *         message: "Empty File"
+ *       }
+ *     ]
+ *   }
+ * };
+ *
+ * try {
+ *   handleCommonErrorResponse(resp, "Error message", "Activity");
+ * } catch (error) {
+ *   console.log(error);
+ * }
+ */
+const handleCommonErrorResponse = (resp, OpErrorMessage, OpActivity, config) => {
+  // checking for invalid/expired token errors and evicting cache in that case
+  // rudderJobMetadata contains some destination info which is being used to evict the cache
+  if (
+    authCache &&
+    resp.response?.errors &&
+    resp.response?.errors?.length > 0 &&
+    resp.response?.errors.some((errorObj) => errorObj.code === '601' || errorObj.code === '602')
+  ) {
```

Comment:
resp is bad name here please lets pick a non confusing name: maybe call `resp` to `apiCallResult`

Reply:
done